src/IR/IR.hs

```diff
@@ -252,3 +255,97 @@ wellFormed (Prim p es _) = wfPrim p es && all wellFormed es
   wfPrim (PrimOp PrimLt    ) [_, _]    = True
   wfPrim (PrimOp PrimLe    ) [_, _]    = True
   wfPrim _                   _         = False
+
+
+{-
+how to use:
+1) stack ghci
+2) import IR.IR
+3) samples:
+show (LitIntegral 3)
+show New
+Questions:
+why does `show (Lit "a" Integer)` not work? -- b/c types not defined, just use () as a type
+why does `show (Var (VarId (Identifier "foo")) LitIntegral)` not work?
+
+show (Var (VarId (Identifier "foo")) ()))
+
+-}
+instance (Pretty t) => Show (Expr t) where
+  show p = show $ pretty p
+
+instance Show Primitive where
+  show p = show $ pretty p
+
+instance Show Literal where
+    show p = show $ pretty p
+
+instance (Pretty t) => Show (Program t) where
+    show p = show $ pretty p
+
+instance Pretty Literal where
+    pretty (LitIntegral i) = pretty (show i)
+    pretty (LitBool b) = pretty (show b)
+
+instance Pretty Primitive where
+  pretty New = pretty "new"
+  pretty Dup = pretty "dup"
+  pretty Drop = pretty "drop"
+  pretty Reuse = pretty "reuse"
+  pretty Deref = pretty "deref"
+  pretty Assign = pretty "="
+  pretty After = pretty "after"
+  pretty Fork = pretty "fork"
+  pretty Wait = pretty "wait"
+  pretty Loop = pretty "loop"
+  pretty Break = pretty "break"
+  pretty Return = pretty "return"
+  pretty (PrimOp p) = pretty "PrimOp" <+> pretty p
+
+instance Pretty PrimOp where
+  pretty PrimNeg = pretty "-"
+  pretty PrimNot = pretty "!"
+  pretty PrimBitNot = pretty "~"
+  pretty PrimAdd = pretty "+"
+  pretty PrimSub = pretty "-"
+  pretty PrimMul = pretty "*"
+  pretty PrimDiv = pretty "/"
+  pretty PrimMod = pretty "%"
+  pretty PrimBitAnd = pretty "&"
+  pretty PrimBitOr = pretty "|"
+  pretty PrimEq = pretty "=="
+  pretty PrimGt = pretty ">"
+  pretty PrimGe = pretty ">="
+  pretty PrimLt = pretty "<"
+  pretty PrimLe = pretty "<="
+
+instance (Pretty t) => Pretty (Expr t) where
+  {- do pretty show(VarId)?
+  this might be needed if other file uses legacy printer
+  var myInt LitIntegral?
+
+  //do things like varId take args?
+
+  //do i have to wrap each expression in parens somehow?
```

Comment:
this is... extremely lispy :eyes: 

You can see the `Pretty` instance for the `Ast` as it is right now for how some of these constructs should look; the AST consists of, roughly speaking, a superset of language constructs found here in the IR.